### ISSUE TITLE
Allow attributes for anonymous records

### ISSUE THREAD

**Issue by leandromoh** [fsharp/fslang-suggestions#1192]
I propose we support attributes for anonymous records.

The existing way of approaching this problem in F# is define a named type.

![image](https://user-images.githubusercontent.com/11452028/192107606-aebc96f8-f970-4b04-9877-f79e6466e43c.png)
_without the attribute the code compiles_

## Pros and Cons 

The advantages of making this adjustment to F# are that we could use anonymous type in more situations that make sense.

I can not see any disadvantages.

## Extra information

Estimated cost: XS or S

**Comment by abelbraaksma** [comment]
Looks like you missed some sections from the issue template….

Anyway, in terms of motivation, have a look here: https://github.com/fsharp/fslang-design/blob/main/FSharp-4.6/FS-1030-anonymous-records.md#design-principle-not-anonymous-object-expressions.

Later in that proposal, you’ll find that specifically `CLIMutable` is considered. This is going to be quite a big change, certainly not XS/S, as it requires careful consideration and implementation of syntax changes (just my 2p, I may be off).

I’m neither against nor in favor if it, I just never had a need for it. As soon as anon types become more complex I’ve always found that making them nominal is the better way out.

**Comment by isaacabraham** [comment]
Funnily enough I came on here tonight to suggest something exactly like this for exactly this reason. Anonymous records are very useful for lightweight, transient values e.g. reading from SQL or JSON into a type that matches the source 1:1, from which you can then do standard F# processing to e.g. create a validated value.

I totally understand the potential reluctance to add this in - slippery slope etc. - but it would a very nice to have.

**Comment by charlesroddie** [comment]
We should move in the opposite direction. If you use a `CLIMutable` you should not only have to define a type, but also write an essay apologizing for the use of this attribute.

**Comment by cartermp** [comment]
Although not called out explicitly, this is likely to fall under the following principle for anonymous records: https://github.com/fsharp/fslang-design/blob/main/FSharp-4.6/FS-1030-anonymous-records.md#design-principle-not-anonymous-object-expressions

I can appreciate that anonymous records being a convenient substitute for a small type declaration that has little meaning beyond its encompassing type sort of falls apart when .NET Interop is concerned. But implementing a capability for the sake of a single use case doesn't feel worth it to me. Are there other attributes scoped to the type declaration that would be handy?

**Comment by isaacabraham** [comment]
> But implementing a capability for the sake of a single use case doesn't feel worth it to me.

Yes, this is my concern, too. It feels very, very specific. But it is a pity without it because the unfortunate truth is that in the .NET world so many serialization / mapping libraries (Dapper, Json etc.) rely on having that ability.

**Comment by dsyme** [comment]
Duplicate of #886.  REsolution is "use a regular record type"
